src/Analyser/NodeScopeResolver.php

```diff
@@ -1890,8 +1890,14 @@ static function (?Type $offsetType, Type $valueType, bool $optional = false) use
 					$setOffsetValueTypes(
 						$scope,
 						$callArgs,
-						static function (?Type $offsetType, Type $valueType) use (&$arrayType): void {
+						static function (?Type $offsetType, Type $valueType, bool $optional) use (&$arrayType): void {
+							$isIterableAtLeastOnce = $arrayType->isIterableAtLeastOnce()->yes() || !$optional;
 							$arrayType = $arrayType->setOffsetValueType($offsetType, $valueType);
+							if ($isIterableAtLeastOnce) {
+								return;
+							}
+
+							$arrayType = TypeUtils::getAnyArrays($arrayType)[0];
```

Comment:
suprised that this is OK, is it? the idea here is to basically remove `NonEmptyArrayType` from a non-empty-array (which is an intersection of ArrayType and NonEmptyArrayType). is there an easier way to do this?